### PR TITLE
Adding lookup plugin for postgresql database

### PR DESCRIPTION
**Pull request by gshivani**
##### SUMMARY
lookup plugin for querying postgresql

##### ISSUE TYPE
Feature Pull Request

##### COMPONENT NAME
lib/ansible/plugins/lookup/postgresql.py

##### ANSIBLE VERSION
2.3

### CODE REVIEW

**Comment by msabramo** [comment]
Maybe even more useful would be a more general plugin that uses SQLAlchemy to query the database. Then you could use it with any of the database engines supported by SQLAlchemy.

http://docs.sqlalchemy.org/en/rel_1_0/dialects/index.html 

Instead of taking username, password, etc., it could take a SQLAlchemy database URL (which probably a lot of people would keep in variables and encrypt with vault).

**Comment by amenonsen** [comment]
A few quick thoughts: (a) this definitely needs documentation, (b) "connection failed or something's wrong" is a complete non-starter, this is something that may well not work right the first time, and that error would drive me nuts while debugging and not help _at all_.

@msabramo's suggestion to use SQLAlchemy is… well, it's probably the Right Thing To Do™, though I personally don't want to have to install and learn SQLAlchemy and don't need anything but Postgres support. Being able to encrypt and store database connection information is certainly nice, but it's also possible to encrypt and store the `conninfo` string for Postgres with this PR already.

File: lib/ansible/plugins/lookup/postgresql.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import psycopg2
```
Comment (bcoca) [inline lib/ansible/plugins/lookup/postgresql.py:20]: this needs to be guarded and return an ansible error in the init  (see consul_kv lookup and HAS_CONSUL)

File: lib/ansible/plugins/lookup/postgresql.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        try:
+            con = psycopg2.connect(database=database, user=user, password=password)
+            cur = con.cursor()
+            cur.execute(sql)
+            rows = cur.fetchall()
+            for row in rows:
+                row_string = str(row)
+                ret.append(row_string)
+        except:
+             raise AnsibleError("Connection Failed or something's wrong")
```
Comment (bcoca) [inline lib/ansible/plugins/lookup/postgresql.py:71]: Adding the exception string to the error would help the user narrow down what the issue is.

File: lib/ansible/plugins/lookup/postgresql.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                assert(name in paramvals)
+                paramvals[name] = value
+        except (ValueError, AssertionError) as e:
+            raise AnsibleError(e)
+        #assigning args for postgresql connect 
+        database = paramvals['database']
+        user = paramvals['user']
+        password = paramvals['password']
+        sql = paramvals['sql']
+        #Connection to postgresql
```
Comment (bcoca) [inline lib/ansible/plugins/lookup/postgresql.py:60]: lacking parameter validation, user should be notified if the minimal requirements for lookup to work are not met

**Review by bcoca — CHANGES_REQUESTED** [review]
I know this has been open a long time, but since previous comments were not addressed we have not returned to it. Now we are trying to close these old issues and if we don't get any response to this review either we'll end up just closing the PR.

**Comment by jstoja** [comment]
@bcoca No answer after +1y, you can probably close it.

**Comment by ryanmerolle** [comment]
What is required here to make this shippable?  I am interested in this becoming a standard lookup plugin in ansible.

File: lib/ansible/plugins/lookup/postgresql.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        #splitting only till 3 parameters because sql query has ' ' inside
+        for term in terms:
+            params = term.split(' ',3)
+
+        paramvals = {
+            'database' : 'database',
+            'user' : 'user',
+            'password' : 'password',
+            'sql' : 'query',
+        }
```
Comment (ryanmerolle) [inline lib/ansible/plugins/lookup/postgresql.py:47]: It likely makes sense to add the remaining psycopg2 parameters including host and port.

**Comment by gundalow** [comment]
@ryanmerolle PR needs rebasing https://docs.ansible.com/ansible/devel/dev_guide/developing_rebasing.html

It maybe worth you creating a new branch and copying across the new file into that branch

**Comment by gshivani** [comment]
@ryanmerolle Glad that you are interested in this becoming a standard plugin! I'm not sure if I can get to this soon. Please feel free to create new branch and reuse the code!

**Comment by Andersson007** [comment]
needs_info

**Pull request closed without merging**